src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -88,26 +106,110 @@ class DocumentStatisticsAnnotationData {
     // ----------------------------------------------------------------------
     using FrequencyMap                      = std::unordered_map<std::string, std::uint32_t>;
     using IndexMap                          = FrequencyMap;
+
+    using FrequencyAndIndexMap              = std::unordered_map<std::string, FrequencyAndIndexStruct>;
     // ----------------------------------------------------------------------
     // |
     // |  Public Data
     // |
     // ----------------------------------------------------------------------
-    FrequencyMap const                      TermFrequency;                  /// Words and the number of documents that it appears in
-    IndexMap const                          TermIndex;
+    FrequencyAndIndexMap const              TermFrequencyAndIndex;
     std::uint32_t const                     TotalNumDocuments;
 
     // ----------------------------------------------------------------------
     // |
     // |  Public Methods
     // |
     // ----------------------------------------------------------------------
-    DocumentStatisticsAnnotationData(FrequencyMap termFreq, IndexMap termIndex, std::uint32_t totalNumDocuments);
+    DocumentStatisticsAnnotationData(FrequencyAndIndexMap termFrequencyAndIndex, std::uint32_t totalNumDocuments);
     ~DocumentStatisticsAnnotationData(void) = default;
 
     FEATURIZER_MOVE_CONSTRUCTOR_ONLY(DocumentStatisticsAnnotationData);
 };
 
+namespace {
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap const & prunedTermFreqByDf, 
+                                                                       std::uint32_t maxFeatures) {
+    std::vector<std::uint32_t> frequencyVector;
```

Comment:
Add the statement `frequencyVector.reserve(prunedTermFreqByDf.size());` so we don't reallocate the vector multiple times as it grows.